src/module.c

```diff
@@ -3604,6 +3638,26 @@ unsigned long long VM_GetClientId(ValkeyModuleCtx *ctx) {
     return ctx->client->id;
 }
 
+/* Sets the client in the current context. Returns VALKEYMODULE_OK on success
+ * and VALKEYMODULE_ERR if there is no client with the provided id.
+ *
+ * This is useful for sending push messages to clients using VM_ReplyWithPush().
+ * It is recommended to restore the client id afterwards as illustrated in the
+ * following example. (Error handling is omitted for brevity.)
+ *
+ *     uint64_t old_id = ValkeyModule_GetClientId(ctx);
+ *     ValkeyModule_SetClientId(ctx, another_client_id);
+ *     ValkeyModule_ReplyWithPush(ctx, 1);
+ *     ValkeyModule_ReplyWithLongLong(ctx, 42);
+ *     ValkeyModule_SetClientId(ctx, old_id);
+ */
+int VM_SetClientId(ValkeyModuleCtx *ctx, uint64_t id) {
```

Comment:
There is something I dislike about making module contexts mutable objects. In all other cases where we acting on a new context, we are generating a new context to operate on, like `ValkeyModule_GetContextFromIO`. I would maybe prefer an API like , `ModuleCtx *ValkeyModule_StartPushReplyToClient(ctx, uint64_t client_id, int length); `

Then we could also at least add validation to make sure there was no misuse of the API, and people do things that the API was intended or tested for.

Reply:
Good idea. Should we require that the context is destroyed after sending the push message? (Then we could set the flag CLIENT_PUSHING that makes it possible to push to self even within multi-exec.)

Reply:
We can't use just the length here to find where the push ends. If the push contains nested arrays and maps, we need a whole stack of lengths and counters to know where the push message ends.

Reply:
> Good idea. Should we require that the context is destroyed after sending the push message? (Then we could set the flag CLIENT_PUSHING that makes it possible to push to self even within multi-exec.)

That sounds reasonable to me.